network/src/connectivity_manager/test.rs

```diff
@@ -1101,3 +1157,75 @@ fn multiple_addrs_shrinking() {
     };
     rt.block_on(f_peer_mgr);
 }
+
+#[test]
+fn test_public_connection_limit() {
+    ::libra_logger::Logger::new().environment_only(true).init();
+    let mut rt = Runtime::new().unwrap();
+    let mut seed_peers: HashMap<PeerId, Vec<NetworkAddress>> = HashMap::new();
+    let mut eligible_peers: HashMap<PeerId, PublicKey> = HashMap::new();
+    for _ in 0..MAX_TEST_CONNECTIONS {
+        let (peer_id, addr, key) = generate_peer_with_addr();
+        eligible_peers.insert(peer_id, key);
+        seed_peers.insert(peer_id, vec![addr]);
+    }
+
+    info!("Seed peers are {:?}", eligible_peers);
+    let network_context = NetworkContext::new(
+        NetworkId::vfn_network(),
+        RoleType::FullNode,
+        PeerId::random(),
+    );
+    let (mut connection_reqs_rx, mut connection_notifs_tx, mut conn_mgr_reqs_tx, mut ticker_tx) =
+        setup_conn_mgr_with_context(network_context, &mut rt, eligible_peers.clone(), seed_peers);
+
+    // Fake peer manager and discovery.
+    let f_peer_mgr = async move {
+        // Peer manager receives a request to connect to the other peer.
+        info!("Waiting to receive dial request");
+        expect_num_dials(
+            &mut connection_reqs_rx,
+            &mut connection_notifs_tx,
+            &mut conn_mgr_reqs_tx,
+            MAX_TEST_CONNECTIONS,
+        )
+        .await;
+
+        // Sending an UpdateEligibleNodes & UpdateAddresses with a new peer should not
+        // trigger any dials.
+        info!("Sending new peer as eligible");
+        let (new_peer_id, new_peer_addr, new_peer_key) = generate_peer_with_addr();
+        eligible_peers.insert(new_peer_id, new_peer_key);
+        conn_mgr_reqs_tx
+            .send(ConnectivityRequest::UpdateEligibleNodes(
+                eligible_peers.clone(),
+            ))
+            .await
+            .unwrap();
+
+        info!("Sending new peer address");
+        conn_mgr_reqs_tx
+            .send(ConnectivityRequest::UpdateAddresses(
+                DiscoverySource::Gossip,
+                [(new_peer_id, vec![new_peer_addr])]
+                    .iter()
+                    .cloned()
+                    .collect(),
+            ))
+            .await
+            .unwrap();
```

Comment:
why go through this headache and not just have them passed in via the new api / seed_peers?

Reply:
True, I could pass in MAX_CONNS + 1 into the seed peers, and there should only be MAX_CONNS.

I wanted to test a 'realistic' example, of having the max connections set up, and then a new connection shows up

Reply:
Yeah, moved to just seed peers to simplify it.